pandas/core/reshape.py

```diff
@@ -558,87 +549,138 @@ def _convert_level_number(level_num, columns):
 
     # this makes life much simpler
-    if level_num != frame.columns.nlevels - 1:
-        # roll levels to put selected level at end
-        roll_columns = this.columns
-        for i in range(level_num, frame.columns.nlevels - 1):
+    # roll levels to put selected level(s) at end
+    roll_columns = this.columns
+    for j, level_num in enumerate(reversed(level_nums)):
+        for i in range(level_num, frame.columns.nlevels - (j + 1)):
             # Need to check if the ints conflict with level names
             lev1 = _convert_level_number(i, roll_columns)
             lev2 = _convert_level_number(i + 1, roll_columns)
             roll_columns = roll_columns.swaplevel(lev1, lev2)
-        this.columns = roll_columns
+    this.columns = roll_columns
 
     if not this.columns.is_lexsorted():
         # Workaround the edge case where 0 is one of the column names,
-        # which interferes with trying to sort based on the first
-        # level
+        # which interferes with trying to sort based on the first level
         level_to_sort = _convert_level_number(0, this.columns)
         this = this.sortlevel(level_to_sort, axis=1)
 
-    # tuple list excluding level for grouping columns
-    if len(frame.columns.levels) > 2:
+    num_levels_to_stack = len(level_nums)
+    level_vals = this.columns.levels[-num_levels_to_stack:]
+    level_labels = sorted(set(zip(*this.columns.labels[-num_levels_to_stack:])))
+    level_vals_used = MultiIndex.from_tuples([tuple(np.nan if lab == -1 else level_vals[i][lab]
+                                                    for i, lab in enumerate(label))
+                                              for label in level_labels],
+                                             names=this.columns.names[-num_levels_to_stack:])
+    levsize = len(level_labels)
+
+    # construct new_index
+    N = len(this)
+    if isinstance(this.index, MultiIndex):
+        new_levels = list(this.index.levels)
+        new_names = list(this.index.names)
+        new_labels = [lab.repeat(levsize) for lab in this.index.labels]
+    else:
+        new_levels = [this.index]
+        new_labels = [np.arange(N).repeat(levsize)]
+        new_names = [this.index.name]  # something better?
+    new_levels += level_vals
+    new_labels += [np.tile(labels, N) for labels in zip(*level_labels)]
+    new_names += level_vals_used.names
+    new_index = MultiIndex(levels=new_levels, labels=new_labels,
+                           names=new_names, verify_integrity=False)
+
+    # if stacking all levels in columns, result will be a Series
+    if len(frame.columns.levels) == num_levels_to_stack:
+        new_data = frame.values.ravel()
+        if dropna:
+            mask = notnull(new_data)
+            new_data = new_data[mask]
+            new_index = new_index[mask]
+        return Series(new_data, index=new_index)
+
+    # result will be a DataFrame
+
+    def _equal_key(keya, keyb):
+        return all(((a == b) or
```

Comment:
I'm trying to get it to pass all tests :-)

Seriously, here's what's happening. I have a DataFrame with ``columns=MutiIndex.from_tuples([('a','w'), ('b','x'), (NaN,'y'), ('d','z')]), and I want to stack the second level. So the result should have ``columns=Index([NaN, 'a', 'b', 'd']). To figure out what goes in the first result column (labeled ``NaN``), I need to find all the columns in the original ``DataFrame`` that begin with ``(NaN,)`` -- i.e. in this case just the third column, labeled ``(NaN,'y')``. How am I supposed to find all such columns? (In this example there is only one, but of course in general there could be many, e.g. if there were a fifth ``(Nan,'z')`` column.) Previously the code used ``columns.get_loc(key)``, but this doesn't work with ``NaN`` in the ``key``. @shoery's suggestion of using ``array_equivalent()`` also doesn't work. And of course there could be a ``NaT`` there instead of a ``NaN``...